lib/sdam/topology.js

```diff
@@ -0,0 +1,416 @@
+
+    const serverDescriptions = seedlist.reduce((result, seed) => {
+      const address = seed.port ? `${seed.host}:${seed.port}` : `${seed.host}:27017`;
+      result.set(address, new ServerDescription(address));
+      return result;
+    }, new Map());
+
+    this.s = {
+      // the id of this topology
+      id: topologyId,
+      // passed in options
+      options: Object.assign({}, options),
+      // initial seedlist of servers to connect to
+      seedlist: seedlist,
+      // the topology description
+      description: new TopologyDescription(
+        topologyType,
+        serverDescriptions,
+        options.replicaset,
+        null,
+        null,
+        options
+      ),
+      serverSelectionTimeoutMS: options.serverSelectionTimeoutMS,
+      heartbeatFrequencyMS: options.heartbeatFrequencyMS,
+      ServerClass: options.ServerClass || null /* eventually our Server class, but null for now */
+    };
+  }
+
+  /**
+   * @return A `TopologyDescription` for this topology
+   */
+  get description() {
+    return this.s.description;
+  }
+
+  /**
+   * Initiate server connect
+   *
+   * @param {Object} [options] Optional settings
+   * @param {Array} [options.auth=null] Array of auth options to apply on connect
+   */
+  connect(/* options */) {
+    // emit SDAM monitoring events
+    this.emit('topologyOpening', new monitoring.TopologyOpeningEvent(this.s.id));
+
+    // emit an event for the topology change
+    this.emit(
+      'topologyDescriptionChanged',
+      new monitoring.TopologyDescriptionChangedEvent(
+        this.s.id,
+        new TopologyDescription(TopologyType.Unknown), // initial is always Unknown
+        this.s.description
+      )
+    );
+
+    // emit `ServerOpeningEvent`s for each server in our topology
+    Array.from(this.s.description.servers.keys()).forEach(serverAddress => {
+      // publish an open event for each ServerDescription created
+      this.emit('serverOpening', new monitoring.ServerOpeningEvent(this.s.id, serverAddress));
+    });
+  }
+
+  /**
+   * Close this topology
+   */
+  close() {
+    // emit an event for close
+    this.emit('topologyClosed', new monitoring.TopologyClosedEvent(this.s.id));
+  }
+
+  /**
+   * Selects a server according to the selection predicate provided
+   *
+   * @param {function} [selector] An optional selector to select servers by, defaults to a random selection within a latency window
+   * @return {Server} An instance of a `Server` meeting the criteria of the predicate provided
+   */
+  selectServer(selector, options, callback) {
+    if (typeof options === 'function') (callback = options), (options = {});
+    options = Object.assign(
+      {},
+      { serverSelectionTimeoutMS: this.s.serverSelectionTimeoutMS },
+      options
+    );
+
+    selectServers(
+      this,
+      selector,
+      options.serverSelectionTimeoutMS,
+      process.hrtime(),
+      (err, servers) => {
+        if (err) return callback(err, null);
```

Comment:
Is there a reason we needed to wrap the `callback` in this arrow function?

Reply:
I don't follow?

Reply:
Ah, I missed `randomSelection`. Nvm.